src/foo/imp.rs

```diff
@@ -209,27 +204,74 @@ pub(crate) mod ffi {
     /// Must be a valid C string, 0-terminated.
     #[no_mangle]
     pub unsafe extern "C" fn ex_foo_new(name: *const c_char) -> *mut ExFoo {
-        let obj = glib::Object::new::<super::super::Foo>(&[(
-            "name",
-            &*glib::GString::from_glib_borrow(name),
-        )])
-        .unwrap();
-        obj.to_glib_full()
+        glib::Object::new::<super::super::Foo>(&[("name", &*glib::GString::from_glib_borrow(name))])
+            .to_glib_full()
     }
 
     #[no_mangle]
     pub extern "C" fn ex_foo_get_type() -> glib::ffi::GType {
         <super::super::Foo as glib::StaticType>::static_type().into_glib()
     }
+
+    #[no_mangle]
+    pub unsafe extern "C" fn ex_foo_check_async(
+        this: *mut ExFoo,
+        cancellable: *mut gio::ffi::GCancellable,
+        callback: gio::ffi::GAsyncReadyCallback,
+        user_data: *mut c_void,
+    ) {
+        let imp = (*this).imp();
+        let obj = &super::super::Foo::from_glib_none(this);
+        let cancellable = gio::Cancellable::from_glib_borrow(cancellable);
+        let callback = callback.unwrap();
+
+        let closure = move |task: gio::LocalTask<bool>, _: Option<&super::super::Foo>| {
+            use glib::Cast;
+
+            let result: *mut gio::ffi::GAsyncResult =
+                task.upcast_ref::<gio::AsyncResult>().to_glib_none().0;
+            callback(this as *mut _, result, user_data)
+        };
+
+        let task = gio::LocalTask::new(Some(obj), Some(&*cancellable), closure);
+
+        glib::MainContext::default().spawn_local(async move {
```

Comment:
```suggestion
        glib::MainContext::thread_default().spawn_local(async move {
```

or not? GIO futures are always spawned on the current thread-default main context

Reply:
That's also wrong in the gtk-rs-core examples apparently:

```
examples/gio_task/file_size/ffi.rs
34:    glib::MainContext::default().spawn_local(async move {

examples/gio_task/file_size/mod.rs
42:        glib::MainContext::default().spawn_local(async move {
```

Reply:
Clearly, I am not knowledgeable enough about this. Rust guarantees that gobjects are Send+Sync?

And it will need something like, to compile:
```
        glib::MainContext::thread_default()
            .unwrap_or_else(|| glib::MainContext::default())
```

Reply:
> Clearly, I am not knowledgeable enough about this. Rust guarantees that gobjects are Send+Sync?

If they are, yes. But that's also irrelevant here :)
The thread default main context is owned by this very thread, so anything running on it will also run on that thread unless ownership of the main context is passed around (which is not really supported and will panic).

In C, GIO async operations also take the thread default main context (if any).

> And it will need something like, to compile:

Or you use this, which does the same :)
https://gtk-rs.org/gtk-rs-core/git/docs/glib/struct.MainContext.html#method.ref_thread_default

Reply:
ok